openvpn-wire-protocol.xml

```diff
@@ -1697,9 +1813,78 @@ authenticated_data_v2 = opcode| key_id | peer_id | packet_id | payload
         </t>
 
         <t>
-          The tag size is always 128 bit (16 bytes). (Same size as in TLS). If the
-          protocol flag for AEAD tag at the end is in effect, the AEAD tag is at the end
-          of the packet rather than at the start.
+          The tag size is always 128 bit (16 bytes). (Same size as in TLS).
+        </t>
+      </section>
+    <section anchor="dataaeadv3" title="AEAD encrypted data channel packet with 64bit counters">
+
+      <t>
+        This is the modern version of the AEAD data format in OpenVPN. The differences to the older
+        format are:
+        <ul>
+          <li>AEAD tag at the end of the packet</li>
+          <li>64 bit packet id </li>
+          <li>IV constructed with XOR instead of contactination </li>
+          <li>Support of multiple key generations (epochs) </li>
+        </ul>
+
+        The packet format with 64bit counter looks like this:
+          <figure>
+            <sourcecode>
+struct aead_packet {
+    int opcode:5;
+    int key_id:3;
+    int peer_id:24;
+    uint_64_t packet_id;
+    uint8_t* encrypted_payload;
+    uint8_t[16] authentication_tag;    
+}
+            </sourcecode>
+          </figure>
+        </t>
+
+        <t>
+          The packet id is split into 16 bit bits of epoch and 48bit of counter for
+          that epoch. Each epoch uses a different a encryption/decryption key and a different
+          implicit_iv.
+          <figure>
+            <sourcecode>
+struct packet_id {
+    uint epoch:16;
+    uint epoch_counter:48;
+}
+            </sourcecode>
+          </figure>
+        </t>
+
+        <t>
+          The epoch_counter of the packet_id together with the implicit IV for the epoch
+          forms the IV for decryption. The epoch_counter is also used for the replay
+          protection.  Implementation are encouraged to a have a per epoch replay protection.
+          Using the whole packet_id as flat 64bit counter for replay protection will work as well
+          but will reject reordered packets from the older epoch when a (valid) packet
+          the epoch changes
```

Comment:
Not sure what to suggest here but the last sentence probably needs slight rephrasing.